mindmeld/app.py

```diff
@@ -161,6 +168,97 @@ def add_dialogue_rule(self, name, handler, **kwargs):
         else:
             self._dialogue_rules.append((name, handler, kwargs))
 
+    def custom_action(self, **kwargs):
+        """Adds a custom action, similar to `add_custom_action` but allows the ordering
+        of nlp entities to be more flexible.
+
+        Examples:
+            app.custom_action(intent='greeting', action='say_greeting')
+            app.custom_action(entity='person', action='greet_person')
+        """
+        action = kwargs.pop("action", None)
+        actions = kwargs.pop("actions", None)
+
+        if not (action or actions):
+            raise CustomActionException(
+                "`action` or `actions` must be present in arguments."
+            )
+
+        if action:
+            self.add_custom_action(
+                action,
+                asynch=kwargs.pop("asynch", False),
+                overwrite=kwargs.pop("overwrite", False),
+                config=kwargs.pop("config", None),
+                **kwargs
+            )
+        else:
+            self.add_custom_action_sequence(
+                actions,
+                asynch=kwargs.pop("asynch", False),
+                overwrite=kwargs.pop("overwrite", False),
+                config=kwargs.pop("config", None),
+                **kwargs
+            )
+
+    def add_custom_action(
+        self, action, asynch=False, overwrite=False, config=None, **kwargs
+    ):
+        """Adds a custom action handler for the dialogue manager.
+
+        Whenever the user hits this state, we invoke the custom action instead and returns
+            the appropriate responder.
+
+        Args:
+            action (str): The name of the custom action
+            asynch (bool): Whether we should invoke this custom action asynchronously
+            overwrite (bool): Whether we should overwrite the Responder with fields from the
```

Comment:
So if overwrite is set of False, and there is a key collision, we just use the key in the original application (and not the custom action)?

Reply:
Mainly, the difference between `overwrite=True` and `overwrite=False` was initially conceived for the directives. Suppose the 3rd party server returns four directives, while the initial responder already has two. With `overwite=True`, we will overwrite the responder's directive field with four new directives. With `overwrite=False`, we will simply append the four new directives to the responder, so we will have a total of six directives.

For the frame, currently, it is a similar idea, when `overwrite=False`, we will "update" the dictionary with new keys and values. If there is a collision, we assume that it is intended by the user since we also send them that `frame` as well.

Reply:
If the directives were passed to the 3rd party service (as we concluded in the design meeting), then the user can simply append/mutate their directives to the existing set of directive, set  `overwrite=True` and expect the DM to pick up these changes. If  `overwrite=False`, I wouldn't expect the `directive` key in the DM to change by the 3rd party service.